source/non_matching/fe_immersed_values.cc

```diff
@@ -0,0 +1,214 @@
+    const unsigned int function_no,
+    const unsigned int quadrature_point) const
+  {
+    const unsigned int component = 0;
+    return shape_surface_grad_component(function_no,
+                                        quadrature_point,
+                                        component);
+  }
+
+
+
+  template <int dim>
+  Tensor<1, dim>
+  FEImmersedSurfaceValues<dim>::shape_surface_grad_component(
+    const unsigned int function_no,
+    const unsigned int quadrature_point,
+    const unsigned int component) const
+  {
+    const Tensor<1, dim> gradient =
+      this->shape_grad_component(function_no, quadrature_point, component);
+    const Tensor<1, dim> &normal = this->normal_vector(quadrature_point);
+
+    return gradient - (normal * gradient) * normal;
+  }
+
+
+
+  template <int dim>
+  const ImmersedSurfaceQuadrature<dim> &
+  FEImmersedSurfaceValues<dim>::get_quadrature() const
+  {
+    return quadrature;
+  }
+
+
+
+  template <int dim>
+  inline void
+  FEImmersedSurfaceValues<dim>::initialize(const UpdateFlags update_flags)
+  {
+    UpdateFlags flags = this->compute_update_flags(update_flags);
+
+    if (flags & (update_JxW_values | update_normal_vectors))
+      flags |= update_covariant_transformation;
+
+    // Initialize the base classes.
+    if (flags & update_mapping)
+      this->mapping_output.initialize(this->n_quadrature_points, flags);
+    this->finite_element_output.initialize(this->n_quadrature_points,
+                                           *this->fe,
+                                           flags);
+
+    // Then get objects into which the FE and the Mapping can store
+    // intermediate data used across calls to reinit. We can do this in
+    // parallel.
+    Threads::Task<
+      std::unique_ptr<typename FiniteElement<dim, dim>::InternalDataBase>>
+      fe_get_data = Threads::new_task(&FiniteElement<dim, dim>::get_data,
+                                      *this->fe,
+                                      flags,
+                                      *this->mapping,
+                                      this->quadrature,
+                                      this->finite_element_output);
+
+    Threads::Task<std::unique_ptr<typename Mapping<dim>::InternalDataBase>>
+      mapping_get_data;
+    if (flags & update_mapping)
+      mapping_get_data = Threads::new_task(&Mapping<dim>::get_data,
+                                           *this->mapping,
+                                           flags,
+                                           this->quadrature);
+
+    this->update_flags = flags;
+
+    // Then collect answers from the two task above.
+    this->fe_data = std::move(fe_get_data.return_value());
+    if (flags & update_mapping)
+      this->mapping_data = std::move(mapping_get_data.return_value());
+    else
+      this->mapping_data =
+        std::make_unique<typename Mapping<dim>::InternalDataBase>();
+  }
+
+
+#include "fe_immersed_values.inst"
```

Comment:
The error looks odd. Maybe you could disable this include if doxygen is built?

Reply:
Yeah, I didn't understand if this was a doxygen bug or something else. I've added your suggestion.

Reply:
Excellent, it worked! I'll have another look at the PR later today!